pimd/pim_nht.c

```diff
@@ -893,10 +893,21 @@ int pim_ecmp_nexthop_lookup(struct pim_instance *pim,
 					nexthop_tab[i].route_metric,
 					nexthop_tab[i].protocol_distance);
 			}
-			/* update nextop data */
+			/* update nexthop data */
 			nexthop->interface = ifp;
-			nexthop->mrib_nexthop_addr =
-				nexthop_tab[i].nexthop_addr;
+			if (PIM_INADDR_IS_ANY(
```

Comment:
For both these added tests, I believe this actually belongs in pim_nexthop:

```
diff --git a/pimd/pim_zlookup.c b/pimd/pim_zlookup.c
index 37e2bfbbf..6c5b51c87 100644
--- a/pimd/pim_zlookup.c
+++ b/pimd/pim_zlookup.c
@@ -217,19 +217,27 @@ static int zclient_read_nexthop(struct pim_instance *pim,
                }
                switch (nexthop_type) {
                case NEXTHOP_TYPE_IFINDEX:
+                       nexthop_tab[num_ifindex].ifindex = stream_getl(s);
+                       nexthop_tab[num_ifindex].protocol_distance = distance;
+                       nexthop_tab[num_ifindex].route_metric = metric;
+                       /*
+                        * Connected route (i.e. no nexthop), use
+                        * address passed in as PIM nexthop.  This will
+                        * allow us to work in cases where we are
+                        * trying to find a route for this box.
+                        */
+                       nexthop_tab[num_ifindex].nexthop_addr.family = AF_INET;
+                       nexthop_tab[num_ifindex].nexthop_addr.prefixlen =
+                               IPV4_MAX_BITLEN;
+                       nexthop_tab[num_ifindex].nexthop_addr.u.prefix4 =
+                               addr;
+                       ++num_ifindex;
+                       break;
                case NEXTHOP_TYPE_IPV4_IFINDEX:
                case NEXTHOP_TYPE_IPV4:
                        nexthop_tab[num_ifindex].nexthop_addr.family = AF_INET;
-                       if (nexthop_type == NEXTHOP_TYPE_IPV4_IFINDEX
-                           || nexthop_type == NEXTHOP_TYPE_IPV4) {
-                               nexthop_tab[num_ifindex]
-                                       .nexthop_addr.u.prefix4.s_addr =
-                                       stream_get_ipv4(s);
-                       } else {
-                               nexthop_tab[num_ifindex]
-                                       .nexthop_addr.u.prefix4.s_addr =
-                                       PIM_NET_INADDR_ANY;
-                       }
+                       nexthop_tab[num_ifindex].nexthop_addr.u.prefix4.s_addr =
+                               stream_get_ipv4(s);
                        nexthop_tab[num_ifindex].ifindex = stream_getl(s);
                        nexthop_tab[num_ifindex].protocol_distance = distance;
                        nexthop_tab[num_ifindex].route_metric = metric;
```

Reply:
This is what I would do.  Can you give it a try and rework the PR if it works?

Reply:
Thanks, I'll give it a try. In fact, I'm currently hunting another issue where my fixed code still fails. Let's see whether your fix solves that as well.

Reply:
There are definitely a couple of other code paths where we look up nexthops that the original code did not work on.  In fact I see a another issue that needs to be addressed( Can you move the setting of nexthop_tab[num_ifindex].protocol_distance and .route_metric to out side the switch statement )?  And remove the default case as well.

Reply:
OK, your fix solved my original issue as well. Unfortunately, the new issue I found today (again resulting in a "can't send join upstream" due to a 0.0.0.0 nexthop from a different code path) still exists. I'll try to have a closer look at this tomorrow.

The fix I used was this:
```
diff --git a/pimd/pim_zlookup.c b/pimd/pim_zlookup.c
index 37e2bfb..9295b23 100644
--- a/pimd/pim_zlookup.c
+++ b/pimd/pim_zlookup.c
@@ -217,22 +217,26 @@ static int zclient_read_nexthop(struct pim_instance *pim,
 		}
 		switch (nexthop_type) {
 		case NEXTHOP_TYPE_IFINDEX:
+			nexthop_tab[num_ifindex].ifindex = stream_getl(s);
+			/*
+			 * Connected route (i.e. no nexthop), use
+			 * address passed in as PIM nexthop.  This will
+			 * allow us to work in cases where we are
+			 * trying to find a route for this box.
+			 */
+			nexthop_tab[num_ifindex].nexthop_addr.family = AF_INET;
+			nexthop_tab[num_ifindex].nexthop_addr.prefixlen =
+				IPV4_MAX_BITLEN;
+			nexthop_tab[num_ifindex].nexthop_addr.u.prefix4 =
+				addr;
+			++num_ifindex;
+			break;
 		case NEXTHOP_TYPE_IPV4_IFINDEX:
 		case NEXTHOP_TYPE_IPV4:
 			nexthop_tab[num_ifindex].nexthop_addr.family = AF_INET;
-			if (nexthop_type == NEXTHOP_TYPE_IPV4_IFINDEX
-			    || nexthop_type == NEXTHOP_TYPE_IPV4) {
-				nexthop_tab[num_ifindex]
-					.nexthop_addr.u.prefix4.s_addr =
-					stream_get_ipv4(s);
-			} else {
-				nexthop_tab[num_ifindex]
-					.nexthop_addr.u.prefix4.s_addr =
-					PIM_NET_INADDR_ANY;
-			}
+			nexthop_tab[num_ifindex].nexthop_addr.u.prefix4.s_addr =
+				stream_get_ipv4(s);
 			nexthop_tab[num_ifindex].ifindex = stream_getl(s);
-			nexthop_tab[num_ifindex].protocol_distance = distance;
-			nexthop_tab[num_ifindex].route_metric = metric;
 			++num_ifindex;
 			break;
 		case NEXTHOP_TYPE_IPV6_IFINDEX:
@@ -272,19 +276,13 @@ static int zclient_read_nexthop(struct pim_instance *pim,
 			}
 			++num_ifindex;
 			break;
-		default:
-			/* do nothing */
-			{
-				char addr_str[INET_ADDRSTRLEN];
-				pim_inet4_dump("<addr?>", addr, addr_str,
-					       sizeof(addr_str));
-				zlog_warn(
-					"%s: found non-ifindex nexthop type=%d for address %s(%s)",
-					__PRETTY_FUNCTION__, nexthop_type,
-					addr_str, pim->vrf->name);
-			}
-			break;
+		case NEXTHOP_TYPE_IPV6:
+		case NEXTHOP_TYPE_BLACKHOLE:
+			/* ignore */
+			continue;
 		}
+		nexthop_tab[num_ifindex].protocol_distance = distance;
+		nexthop_tab[num_ifindex].route_metric = metric;
 	}
 
 	return num_ifindex;
```
So your version with moved .protocol_distance and .route_metric assignments as requested. Completely removing the default case didn't work as the compiler then complained about unhandled enum values. Instead, I listed the remaining values explicitly and ignored them silently, but I'm not 100% sure that's the right thing to do.

Reply:
Found it. pim_parse_nexthop_update() needed a similar special treatment of NEXTHOP_TYPE_IFINDEX. I'll update my pull request in a few minutes.